### ISSUE TITLE
channel types and other typing problems

### ISSUE THREAD

**Issue by floren** [mattn/anko#179]
Apologies if this veers a little... philosophical.

Anko comprehends a bunch of types, but the way things are implemented means you don't have to think about types until you REALLY NEED TO think about types, and then you're stuck.

I can define a channel of bytes:
```
> c = make(chan byte, 1)
(chan uint8)(0xc420070070)
```
Having done that, one might think I could do this:
```
> c <- 10
panic: reflect.Value.Send: value of type int64 is not assignable to type uint8
```
Instead, the only thing I've been able to figure out is:
```
> a = make([]byte, 1)
[]byte{0x0}
> a[0] = 0xa
10
> c <- a[0]
<nil>
> <-c
0xa
```

I thought I might be able to trick it, but no:
```
> a = make(byte)
0x0
> typeOf(a)
"uint8"
> a = 10
10
> typeOf(a)
"int64"
```
Basically, this issue is me griping about inconsistencies in Anko's handling of types. You can define typed arrays and channels using `make`, but you cannot define a scalar or map of a particular type.

Of course, this might go completely against what @mattn intends for the project, and I'll accept that. In that case, though, it feels like all arrays should be `[]interface` and all chans should be channels of `interface`.

**Comment by MichaelS11** [comment]
In general, not just this case, I would like all the type conversions to be handled, or at least a much as possible. I have been updating the code to handle more of them but channels is low on the list for me. Right now I am still adding testing and bug fixes for function conversions. After that I need to revisit and refactor slice conversions. Also on my list is adding syntax like make(map int64, string, 3). Lots left to do, welcome help and good PRs. :)

Also note you can define any types you want, including map and chan, outside of the program with DefineType

https://godoc.org/github.com/mattn/anko/vm#Env.DefineType

**Comment by MichaelS11** [comment]
A possible work around:

```
a = make([]byte)
a += 10
b = typeOf(a[0])
println(b)
# uint8
```

**Comment by floren** [comment]
Thanks @MichaelS11 for the rapid response. The reason I'm running into these sort of edge cases is that I'm writing documentation, so I'm trying to explore the capabilities of things like channels and maps to get them noted down. At this point, I'm planning to write something to the effect of "be very careful about defining channels of anything except interface, float64, int64, string, or bool".

Regarding your workaround in the second comment, yeah that's what basically what I was doing here:
```
> a = make([]byte, 1)
[]byte{0x0}
> a[0] = 0xa
10
> c <- a[0]
<nil>
> <-c
0xa
```

**Comment by MichaelS11** [comment]
Thanks for your help :)

My bad for retyping your work around, missed it, read too fast.

**Comment by mattn** [comment]
How about convert to the type of chan?

```diff
diff --git a/vm/vmExpr.go b/vm/vmExpr.go
index 46c33ad..c9408cc 100644
--- a/vm/vmExpr.go
+++ b/vm/vmExpr.go
@@ -681,6 +681,10 @@ func invokeExpr(expr ast.Expr, env *Env) (reflect.Value, error) {
 				return nilValue, newError(e.Lhs, err)
 			}
 			if lhs.Kind() == reflect.Chan {
+				lt := lhs.Type().Elem()
+				if rhs.Type().ConvertibleTo(lt) {
+					rhs = rhs.Convert(lt)
+				}
 				lhs.Send(rhs)
 				return nilValue, nil
 			} else if rhs.Kind() == reflect.Chan {
```

**Comment by MichaelS11** [comment]
Yeah, something like that. Was thinking of using the new convert function though.

**Comment by floren** [comment]
Seems sensible to me.

**Comment by mattn** [comment]
@MichaelS11 If you don't have opinion to my change above, I'll fix as so.

**Comment by MichaelS11** [comment]
That works. Was saying you could also see about using convertReflectValueToType instead.

The general idea like:

if lhs.Kind() == reflect.Chan {
 // blah
} else {
convertReflectValueToType 
// blah
} 

Somewhat similar to:
https://github.com/mattn/anko/blob/master/vm/vmExpr.go#L725-L730

**Comment by mattn** [comment]
@MichaelS11 Could you please send PR? :)

**Comment by MichaelS11** [comment]
Sure :)

**Comment by MichaelS11** [comment]
https://github.com/mattn/anko/pull/181

**Comment by MichaelS11** [comment]
Close this?

**Comment by MichaelS11** [comment]
@mattn Close this?
